sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+                DataAccumulatorHook(synapse, data_dir, synapse.get_loggable(), steps)
+        else:
+            for comp in args:
+                DataAccumulatorHook(comp, data_dir, comp.get_loggable(), steps)
+
+    # To micromanage the forward/backward sweeps, subclass Network and override summation(), forward(), backward().
+    @staticmethod
+    def summation(synaptic_input: list[torch.tensor]) -> torch.tensor:
+        """Adds up inputs from multiple synapse objects onto the same ensemble, given as rows.
+
+        Note
+        ----
+        If your model requires some preprocessing of inputs to the postsynaptic neuron, it can be implemented
+        by overriding this method.
+
+        """
+        return torch.sum(torch.vstack(synaptic_input), dim=0)
+
+    def backward(self) -> None:
+        """Processes a backward sweep for this network object.
+
+        This is not required for SNNs that learn with STDP and can be implemented by child classes that
+        might require a backward pass.
+
+        """
+        pass
+
+    def forward(self, data: torch.tensor) -> dict:
+        """Processes current simulation step for this network object.
+
+        In this generic implementation, forward call order is determined by a BFS traversal starting from
+        the edges fanning out of the root nodes. Every node is visited once; for each node, every outgoing synapse
+        is forwarded. That way, we are guaranteed to forward any component exactly once on every simulation step.
+
+        Parameters
+        ----------
+        data: torch.tensor
+            External input to be processed by this generic network.
+
+        Returns
+        -------
+        dict
+            Dictionary containing the loggable properties of this network's ensembles and synapses in this
+            simulation step.
+
+        Warning
+        -------
+        Networks now support multiple root nodes and external input streams. In those cases, the graph is
+        traversed layer-wise, using multi-source BFS.
+
+        See Also
+        --------
+        :meth:`simulation.Simulator.run`
+            For a demonstration of how to programmatically feed heterogeneous external current to multiple nodes.
+            Note that this functionality will eventually be relegated to the data loader, with arbitrary current
+            injections during the simulation being implemented by biases.
+
+        """
+        # follow outgoing synapses between vertices, forwarding the entire network on each iteration.
+        for i, ensemble in enumerate(self.traversal_order):
+            # list all synapses coming in and out of this ensemble. Networkx references it by its string identifier.
+            incoming_synapses = self._in_edges(ensemble)
+            outgoing_synapses = self._out_edges(ensemble)
+
+            # shortcut to actual ensemble reference.
+            ensemble_ref = self.graph.nodes[ensemble].get("reference")
```

Comment:
For the same reasons I don't like warnings :) this should be `self.graph.nodes[ensemble]["reference"]`. Because either the property should exist and not be None or raise a key error. Because down below you're not handling a null case. So seeing a get, which can return a null makes me nervous.